### PR TITLE
Improve error handling and messaging in injection_napari_dependencies hints

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Arg... I'm so mad, I just wrote a long detailed explanation for this and then somehow lost it before i could submit :angry:  ... so now you get the short version :/

This PR came about because, on main, if you run the following code:

```python
@inject_napari_dependencies
def func(param: 'UnknownType' = 1): ...
```

you'll get `NameError: name 'UnknownType' is not defined`.

However, since it has a default, it really isn't necessary to resolve it.

So this PR:
- breaks out the type resolution logic into an independent, reusable module, with lots of new tests
- adds lots of docstrings and inline comments about how it all works
- catches a number of potential "gotchas" when injecting dependencies into a function (for example, a function with a required parameter that had _no_ type hints would have previously made it through without warning, but then fail when called later).
- is more forgiving in the cases above, when a default parameter is already provided
- adds a lot of informative errors and/or warnings (all of which can be suppressed or raised depending on parameters)
- brings test coverage on the `_injection` module close to 100%

-----

some additional context here:  I see `inject_dependencies` as the best way forward to use "regular old" well-annotated python functions to provide context specific functionality (such as acting on the selected layer, getting a viewer, etc.).  I see it as something that both plugin functions will go through as well as internal functions.  So it's worth making it very solid and clear

### CODE REVIEW

**Comment by Czaki** [comment]
Looks ok. 
I have one question (that does not need to be resolved in this PR). If someone will use inject dependency on the method of qt widget form plugin it may keep it alive even after the close of the widget if it is connected to some signal?

**Comment by tlambert03** [comment]
> If someone will use inject dependency on the method of qt widget form plugin it may keep it alive even after the close of the widget if it is connected to some signal?

a very good question.  let me look into it

File: napari/utils/_injection/_type_resolution.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        'return': <class 'napari.layers.labels.labels.Labels'>
+    }
+    """
+    _localns = dict(_typing_names())
+    if inject_napari_namespace:
+        _localns.update(_napari_names())
+    if localns:
+        _localns.update(localns)  # explicitly provided locals take precedence
+    return typing.get_type_hints(
+        obj, globalns=globalns, localns=_localns, include_extras=include_extras
```
Comment (Carreau) [inline napari/utils/_injection/_type_resolution.py]: `include_extras` is 3.9+ only. and I believe is the source of tests failures.
... we could drop 3.8, but I believe this is a bit early to do so.

Comment (tlambert03) [reply napari/utils/_injection/_type_resolution.py]: i put the fix in the wrong function :joy: will move

**Comment by tlambert03** [comment]
i'm actually thinking about pulling this out into a new package.  It's rather generic.  (the only part about this that is specific to napari is a single dictionary that injects names).  That would allow easier performance scrutiny/benchmarks, test iteration, etc...  working on it here: https://github.com/tlambert03/in-n-out

(name can change :joy: partially in honor of our hackathon, but also cause this processes ins and outs :joy:)

**Comment by Carreau** [comment]
Do you want to wait until you have it sorted out in it's own package ?

**Comment by tlambert03** [comment]
> Do you want to wait until you have it sorted out in it's own package ?

well, I'll leave it up to you and @napari/core-devs.  Since this is strictly an improvement over the existing behavior (aims to have no breakages, supports more use cases, and provides more useful feedback when it fails), this could be merged while I work on the other package ... which is coming along. It would be useful for some of the other plugins/commands PRs

But, if the burden of review feels too big, that's fine too.

**Comment by tlambert03** [comment]
since there's been no response, and since this is included in in-n-out (which is included in #4784), I'm going to close this.  but this _is_ a strict improvement over what we currently have, so if for any reason #4784 doesn't get accepted, this should be reopened and merged

**Pull request closed without merging**